logging middleware: log grpc metadata, including user-agent

jrockway opened the pull request:
This helps with `pachyderm-sdk`'s desire to send some headers with each request, for client version number and whatnot.

`log.Metadata` already knows how to redact the auth token, and does.

robert-uhl reviewed (changes requested):
I don’t see anything in `log.Metadata` which sanitizes the auth token; I do see that there is a `log.OutgoingMetadata` field type; perhaps there should be a corresponding `log.IncomingMetadata`, or perhaps `log.Metadata` could detect `constants.ContextTokenKey`?

jrockway commented:
@robert-uhl Good catch.  I've adjusted Metadata to handle removing the auth token (and changed it to "[MASKED]" which is what protoc-gen-zap does).  PTAL.

robert-uhl commented on src/internal/log/field.go:
```diff
@@ -65,14 +65,17 @@ func RetryAttempt(i int, max int) Field {
 			v := md[k]
+			if k == constants.ContextTokenKey {
+				v = []string{"[MASKED]"}
+			}
```
Right now it’s only possible to have a single token, so this is okay. It _might_ be worth iterating through the value, setting each item to `"MASKED"`, in order to preserve an a congruent array, in case zero or multiple tokens someday make it into a request.
